zb-db/src/main/java/io/camunda/zeebe/db/impl/rocksdb/transaction/TransactionalColumnFamily.java

```diff
@@ -94,160 +124,152 @@ private DirectBuffer getValue(
     ensureInOpenTransaction(
-        context,
         transaction -> {
           columnFamilyContext.writeKey(key);
-          getValue(context, columnFamilyContext);
+          final byte[] value =
+              transaction.get(
+                  transactionDb.getDefaultNativeHandle(),
+                  transactionDb.getReadOptionsNativeHandle(),
+                  columnFamilyContext.getKeyBufferArray(),
+                  columnFamilyContext.getKeyLength());
+          columnFamilyContext.wrapValueView(value);
         });
     return !columnFamilyContext.isValueViewEmpty();
   }
 
   @Override
   public boolean isEmpty() {
     final AtomicBoolean isEmpty = new AtomicBoolean(true);
-    whileEqualPrefix(
-        context,
-        keyInstance,
-        valueInstance,
-        (key, value) -> {
-          isEmpty.set(false);
-          return false;
-        });
-    return isEmpty.get();
-  }
-
-  public void forEach(final TransactionContext context, final Consumer<ValueType> consumer) {
-    whileEqualPrefix(
-        context,
-        keyInstance,
-        valueInstance,
-        (BiConsumer<KeyType, ValueType>) (ignore, value) -> consumer.accept(value));
-  }
+    ensureInOpenTransaction(
+        transaction ->
+            forEachInPrefix(
+                new DbNullKey(),
+                (key, value) -> {
+                  isEmpty.set(false);
+                  return false;
+                }));
 
-  public void forEach(
-      final TransactionContext context, final BiConsumer<KeyType, ValueType> consumer) {
-    whileEqualPrefix(context, keyInstance, valueInstance, consumer);
+    return isEmpty.get();
   }
 
-  public void whileTrue(
-      final TransactionContext context, final KeyValuePairVisitor<KeyType, ValueType> visitor) {
-    whileEqualPrefix(context, keyInstance, valueInstance, visitor);
+  private void assertDoesNotExist(final ZeebeTransaction transaction) throws Exception {
+    final var value =
+        transaction.get(
+            transactionDb.getDefaultNativeHandle(),
+            transactionDb.getReadOptionsNativeHandle(),
+            columnFamilyContext.getKeyBufferArray(),
+            columnFamilyContext.getKeyLength());
+    if (value != null) {
+      throw new IllegalStateException("Key already exists");
+    }
   }
 
-  public void whileEqualPrefix(
-      final TransactionContext context,
-      final DbKey keyPrefix,
-      final BiConsumer<KeyType, ValueType> visitor) {
-    whileEqualPrefix(context, keyPrefix, keyInstance, valueInstance, visitor);
+  private void assertExists(final ZeebeTransaction transaction) throws Exception {
+    final var value =
+        transaction.get(
+            transactionDb.getDefaultNativeHandle(),
+            transactionDb.getReadOptionsNativeHandle(),
+            columnFamilyContext.getKeyBufferArray(),
+            columnFamilyContext.getKeyLength());
+    if (value == null) {
+      throw new IllegalStateException("Key does not exist");
```

Comment:
Just as a comment, it would be nice if we add more context to exceptions like this. For example the key, ideally we add as message what we expected at this case and what the actual state is (plus context) see also https://github.com/camunda-cloud/zeebe/wiki/Error-Guidelines. Otherwise the exception is not really helpful and we have to spent a long time to find out which key caused that etc.